File: CBLClient/Apps/CBLTestServer-Android/app/src/main/java/com/couchbase/CouchbaseLiteServ/server/RequestHandler/ExpressionRequestHandler.java
package com.couchbase.CouchbaseLiteServ.server.RequestHandler;


import com.couchbase.CouchbaseLiteServ.server.Args;
import com.couchbase.lite.ArrayExpression;
import com.couchbase.lite.ArrayExpressionIn;
import com.couchbase.lite.Expression;
import com.couchbase.lite.Meta;
import com.couchbase.lite.MetaExpression;
import com.couchbase.lite.PropertyExpression;
import com.couchbase.lite.VariableExpression;


public class ExpressionRequestHandler {

    public PropertyExpression property(Args args) {
        String property = args.get("property");
        return Expression.property(property);
    }

    public MetaExpression metaId() {
        return Meta.id;
    }

    public MetaExpression metaSequence() {
        return Meta.sequence;
    }

    public Expression parameter(Args args) {
        String parameter = args.get("parameter");
        return Expression.parameter(parameter);
    }

    public Expression negated(Args args) {
        Expression expression = args.get("expression");
        return Expression.negated(expression);
    }


    public Expression not(Args args) {
        Expression expression = args.get("expression");
        return Expression.not(expression);
    }

    public VariableExpression variable(Args args) {
        String name = args.get("name");
        return ArrayExpression.variable(name);
    }


    public ArrayExpressionIn any(Args args) {
        VariableExpression variable = args.get("variable");
        return ArrayExpression.any(variable);
    }


    public ArrayExpressionIn anyAndEvery(Args args) {
        VariableExpression variable = args.get("variable");
        return ArrayExpression.anyAndEvery(variable);
    }


    public ArrayExpressionIn every(Args args) {
        VariableExpression variable = args.get("variable");
        return ArrayExpression.every(variable);
    }


    public Expression createEqualTo(Args args) {
        Expression expression1 = args.get("expression1");
        Expression expression2 = args.get("expression2");
        return expression1.equalTo(expression2);
    }


    public Expression createAnd(Args args) {
        Expression expression1 = args.get("expression1");
        Expression expression2 = args.get("expression2");
        return expression1.and(expression2);
    }

    public Expression createOr(Args args) {
        Expression expression1 = args.get("expression1");
        Expression expression2 = args.get("expression2");
        return expression1.or(expression2);
    }

}

File: CBLClient/Apps/CBLTestServer-iOS/CBLTestServer-iOS/Server/ExpressionRequestHandler.swift
//
//  ExpressionRequestHandler.swift
//  CBLTestServer-iOS
//
//  Created by Raghu Sarangapani on 1/16/18.
//  Copyright © 2018 Raghu Sarangapani. All rights reserved.
//

import Foundation
import CouchbaseLiteSwift


public class ExpressionRequestHandler {
    public static let VOID: String? = nil
    fileprivate var _pushPullReplListener:NSObjectProtocol?
    
    public func handleRequest(method: String, args: Args) throws -> Any? {
        switch method {
        ////////////////
        // Expression //
        ////////////////
        case "expression_property":
            let property: String = args.get(name: "property")!
            return Expression.property(property)

        case "expression_metaId":
            return Meta.id
            
        case "expression_metaSequence":
            return Meta.sequence

        case "expression_parameter":
            let parameter: String = args.get(name: "parameter")!
            return Expression.parameter(parameter)
            
        case "expression_negated":
            let expression: Expression = args.get(name: "expression")!
            return Expression.negated(expression as! ExpressionProtocol)
            
        case "expression_not":
            let expression: Expression = args.get(name: "expression")!
            return Expression.not(expression as! ExpressionProtocol)
        
        case "expression_createEqualTo":
            let expression1: ExpressionProtocol = args.get(name: "expression1")!
            let expression2: ExpressionProtocol = args.get(name: "expression2")!
            return expression1.equalTo(expression2)
            
        case "expression_createAnd":
            let expression1: ExpressionProtocol = args.get(name: "expression1")!
            let expression2: ExpressionProtocol = args.get(name: "expression2")!
            return expression1.and(expression2)
            
        case "expression_createOr":
            let expression1: ExpressionProtocol = args.get(name: "expression1")!
            let expression2: ExpressionProtocol = args.get(name: "expression2")!
            return expression1.or(expression2)
        
        default:
            throw RequestHandlerError.MethodNotFound(method)
        }
    }
}
